File: vaadin-nvl/src/main/resources/META-INF/frontend/nvl-graph.ts
import { html, LitElement } from "lit";
import { customElement, property } from "lit/decorators.js";
import NVL from "@neo4j-nvl/base";
import type {
  Node,
  Relationship,
  NvlOptions,
  Layout,
  LayoutOptions,
  ZoomOptions,
} from "@neo4j-nvl/base";
import {
  PanInteraction,
  ZoomInteraction,
  DragNodeInteraction,
  ClickInteraction,
} from "@neo4j-nvl/interaction-handlers";

@customElement("nvl-graph")
export class NvlGraph extends LitElement {
  private nvlInstance: NVL | null = null;
  private container: HTMLDivElement | null = null;

  // Interaction handlers
  private panInteraction: PanInteraction | null = null;
  private zoomInteraction: ZoomInteraction | null = null;
  private dragNodeInteraction: DragNodeInteraction | null = null;
  private clickInteraction: ClickInteraction | null = null;

  // Popup state
  private popupOverlay: HTMLDivElement | null = null;
  private popupTargetId: string | null = null;
  private popupTargetIsRelationship = false;
  private popupTrackingRaf: number | null = null;

  @property({ type: Array })
  nodes: Node[] = [];

  @property({ type: Array })
  relationships: Relationship[] = [];

  @property({ type: Object })
  options: NvlOptions = {};

  @property({ type: Boolean })
  nodeDraggingEnabled = false;

  // Disable shadow DOM so NVL gets normal DOM event handling
  createRenderRoot() {
    return this;
  }

  render() {
    return html`
      <style>
        nvl-graph {
          display: block;
          width: 100%;
          height: 100%;
          position: relative;
        }
        .nvl-container {
          width: 100%;
          height: 100%;
          min-height: 400px;
        }
        .nvl-popup-overlay {
          position: absolute;
          z-index: 100;
          pointer-events: auto;
          display: none;
        }
      </style>
      <div class="nvl-container"></div>
    `;
  }

  firstUpdated() {
    this.container = this.querySelector(".nvl-container") as HTMLDivElement;
    // Create popup overlay imperatively (outside Lit's template) so Lit
    // re-renders never wipe its children.
    this.popupOverlay = document.createElement("div");
    this.popupOverlay.classList.add("nvl-popup-overlay");
    this.appendChild(this.popupOverlay);
    this.initNvl();
  }

  updated(changedProperties: Map<string, unknown>) {
    if (
      changedProperties.has("nodes") ||
      changedProperties.has("relationships")
    ) {
      if (this.nvlInstance) {
        this.nvlInstance.addAndUpdateElementsInGraph(
          this.nodes,
          this.relationships
        );
      } else {
        this.initNvl();
      }
    }
    if (changedProperties.has("nodeDraggingEnabled")) {
      this.updateDragInteraction();
    }
  }

  private initNvl() {
    if (!this.container) return;
    this.destroyNvl();

    const opts: NvlOptions = {
      initialZoom: 1,
      renderer: "canvas",
      ...this.options,
    };

    this.nvlInstance = new NVL(
      this.container,
      this.nodes,
      this.relationships,
      opts,
      {
        onLayoutDone: () => this.fireSimpleEvent("nvl-layout-done"),
        onLayoutComputing: (isComputing: boolean) =>
          this.dispatchEvent(
            new CustomEvent("nvl-layout-computing", {
              detail: { isComputing },
              bubbles: true,
              composed: true,
            })
          ),
        onZoomTransitionDone: () =>
          this.fireSimpleEvent("nvl-zoom-transition-done"),
        onInitialization: () => this.fireSimpleEvent("nvl-initialization"),
      }
    );

    // Set up interaction handlers
    this.panInteraction = new PanInteraction(this.nvlInstance);

    this.zoomInteraction = new ZoomInteraction(this.nvlInstance);

    this.clickInteraction = new ClickInteraction(this.nvlInstance);
    this.clickInteraction.updateCallback("onNodeClick", (node, hitElements, event) => {
      this.fireHitEvent("nvl-click", hitElements);
    });
    this.clickInteraction.updateCallback("onRelationshipClick", (rel, hitElements, event) => {
      this.fireHitEvent("nvl-click", hitElements);
    });
    this.clickInteraction.updateCallback("onCanvasClick", (_event) => {
      this.dispatchEvent(
        new CustomEvent("nvl-click", {
          detail: { nodeIds: [], relationshipIds: [] },
          bubbles: true,
          composed: true,
        })
      );
    });
    this.clickInteraction.updateCallback("onNodeDoubleClick", (node, hitElements, event) => {
      this.fireHitEvent("nvl-dblclick", hitElements);
    });
    this.clickInteraction.updateCallback("onRelationshipDoubleClick", (rel, hitElements, event) => {
      this.fireHitEvent("nvl-dblclick", hitElements);
    });
    this.clickInteraction.updateCallback("onCanvasDoubleClick", (_event) => {
      this.dispatchEvent(
        new CustomEvent("nvl-dblclick", {
          detail: { nodeIds: [], relationshipIds: [] },
          bubbles: true,
          composed: true,
        })
      );
    });
    this.clickInteraction.updateCallback("onNodeRightClick", (node, hitElements, event) => {
      this.fireHitEvent("nvl-contextmenu", hitElements);
    });
    this.clickInteraction.updateCallback("onRelationshipRightClick", (rel, hitElements, event) => {
      this.fireHitEvent("nvl-contextmenu", hitElements);
    });
    this.clickInteraction.updateCallback("onCanvasRightClick", (_event) => {
      this.dispatchEvent(
        new CustomEvent("nvl-contextmenu", {
          detail: { nodeIds: [], relationshipIds: [] },
          bubbles: true,
          composed: true,
        })
      );
    });

    this.updateDragInteraction();
  }

  private updateDragInteraction() {
    if (!this.nvlInstance) return;

    if (this.nodeDraggingEnabled && !this.dragNodeInteraction) {
      this.dragNodeInteraction = new DragNodeInteraction(this.nvlInstance);
      this.dragNodeInteraction.updateCallback("onDragEnd", (nodes: Node[], _evt: MouseEvent) => {
        for (const node of nodes) {
          const pos = this.nvlInstance!.getPositionById(node.id);
          this.dispatchEvent(
            new CustomEvent("nvl-node-drag-end", {
              detail: {
                nodeId: node.id,
                x: pos?.x ?? 0,
                y: pos?.y ?? 0,
              },
              bubbles: true,
              composed: true,
            })
          );
        }
      });
      // Don't let pan fight with node drag
      this.panInteraction?.updateTargets(["node", "relationship"], false);
    } else if (!this.nodeDraggingEnabled && this.dragNodeInteraction) {
      this.dragNodeInteraction.destroy();
      this.dragNodeInteraction = null;
      this.panInteraction?.updateTargets([], true);
    }
  }

  // ===== Event helpers =====

  private fireHitEvent(
    eventName: string,
    hitTargets: { nodes: Array<{ data: { id: string } }>; relationships: Array<{ data: { id: string } }> }
  ) {
    this.dispatchEvent(
      new CustomEvent(eventName, {
        detail: {
          nodeIds: hitTargets.nodes.map((n) => n.data.id),
          relationshipIds: hitTargets.relationships.map((r) => r.data.id),
        },
        bubbles: true,
        composed: true,
      })
    );
  }

  private fireSimpleEvent(name: string) {
    this.dispatchEvent(
      new CustomEvent(name, { bubbles: true, composed: true })
    );
  }

  // ===== Graph data methods =====

  addAndUpdateElements(nodes: Node[], relationships: Relationship[]) {
    this.nvlInstance?.addAndUpdateElementsInGraph(nodes, relationships);
  }

  updateElements(nodes: Node[], relationships: Relationship[]) {
    this.nvlInstance?.updateElementsInGraph(nodes, relationships);
  }

  addElements(nodes: Node[], relationships: Relationship[]) {
    this.nvlInstance?.addElementsToGraph(nodes, relationships);
  }

  removeNodes(nodeIds: string[]) {
    this.nvlInstance?.removeNodesWithIds(nodeIds);
  }

  removeRelationships(relationshipIds: string[]) {
    this.nvlInstance?.removeRelationshipsWithIds(relationshipIds);
  }

  // ===== Layout methods =====

  setLayout(layout: string) {
    this.nvlInstance?.setLayout(layout as Layout);
  }

  setLayoutOptions(options: LayoutOptions) {
    this.nvlInstance?.setLayoutOptions(options);
  }

  // ===== Zoom & pan methods =====

  setZoom(zoomValue: number) {
    this.nvlInstance?.setZoom(zoomValue);
  }

  setPan(panX: number, panY: number) {
    this.nvlInstance?.setPan(panX, panY);
  }

  setZoomAndPan(zoom: number, panX: number, panY: number) {
    this.nvlInstance?.setZoomAndPan(zoom, panX, panY);
  }

  resetZoom() {
    this.nvlInstance?.resetZoom();
  }

  fit(nodeIds: string[], zoomOptions?: ZoomOptions) {
    this.nvlInstance?.fit(nodeIds, zoomOptions);
  }

  fitAll(zoomOptions?: ZoomOptions) {
    if (!this.nvlInstance) return;
    const nodes = this.nvlInstance.getNodes();
    this.nvlInstance.fit(
      nodes.map((n) => n.id),
      zoomOptions
    );
  }

  // ===== Selection methods =====

  deselectAll() {
    this.nvlInstance?.deselectAll();
  }

  // ===== Pinning methods =====

  pinNode(nodeId: string) {
    this.nvlInstance?.pinNode(nodeId);
  }

  unpinNodes(nodeIds: string[]) {
    this.nvlInstance?.unPinNode(nodeIds);
  }

  // ===== Node positions =====

  setNodePositions(data: Node[], updateLayout: boolean) {
    this.nvlInstance?.setNodePositions(data, updateLayout);
  }

  // ===== Renderer =====

  setRenderer(renderer: string) {
    this.nvlInstance?.setRenderer(renderer);
  }

  // ===== Restart =====

  restart(options: NvlOptions, retainPositions: boolean) {
    this.nvlInstance?.restart(options, retainPositions);
  }

  // ===== Export methods =====

  saveToFile(options: { filename?: string; backgroundColor?: string }) {
    this.nvlInstance?.saveToFile(options);
  }

  saveToSvg(options: { filename?: string; backgroundColor?: string }) {
    this.nvlInstance?.saveToSvg(options);
  }

  saveFullGraphToLargeFile(options: {
    filename?: string;
    backgroundColor?: string;
  }) {
    this.nvlInstance?.saveFullGraphToLargeFile(options);
  }

  // ===== Getter methods (return values for server round-trip) =====

  getScale(): number {
    return this.nvlInstance?.getScale() ?? 1;
  }

  getPan(): { x: number; y: number } {
    return this.nvlInstance?.getPan() ?? { x: 0, y: 0 };
  }

  getZoomLimits(): { minZoom: number; maxZoom: number } {
    return (
      this.nvlInstance?.getZoomLimits() ?? { minZoom: 0.075, maxZoom: 10 }
    );
  }

  getNodes(): Node[] {
    return this.nvlInstance?.getNodes() ?? [];
  }

  getRelationships(): Relationship[] {
    return this.nvlInstance?.getRelationships() ?? [];
  }

  getSelectedNodes(): Node[] {
    return this.nvlInstance?.getSelectedNodes() ?? [];
  }

  getSelectedRelationships(): Relationship[] {
    return this.nvlInstance?.getSelectedRelationships() ?? [];
  }

  getNodePositions(): Node[] {
    return this.nvlInstance?.getNodePositions() ?? [];
  }

  isLayoutMoving(): boolean {
    return this.nvlInstance?.isLayoutMoving() ?? false;
  }

  getImageDataUrl(options: { backgroundColor?: string }): string {
    return this.nvlInstance?.getImageDataUrl(options) ?? "";
  }

  // ===== Popup methods =====

  showPopup(elementId: string, isRelationship: boolean) {
    this.popupTargetId = elementId;
    this.popupTargetIsRelationship = isRelationship;
    if (this.popupOverlay) {
      // Move all server-appended popup content (direct children that aren't
      // the container, overlay, or style elements) into the overlay div.
      const contentChildren = Array.from(this.children).filter(
        (child) =>
          child !== this.popupOverlay &&
          child !== this.container &&
          child.tagName !== "STYLE"
      );
      for (const child of contentChildren) {
        this.popupOverlay.appendChild(child);
      }
      this.popupOverlay.style.display = "block";
    }
    this.updatePopupPosition();
    this.startPopupTracking();
  }

  hidePopup() {
    this.stopPopupTracking();
    this.popupTargetId = null;
    if (this.popupOverlay) {
      this.popupOverlay.style.display = "none";
      // Move children back out of the overlay so Vaadin can remove them
      while (this.popupOverlay.firstChild) {
        this.appendChild(this.popupOverlay.firstChild);
      }
    }
  }

  private getWorldPosition(
    elementId: string,
    isRelationship: boolean
  ): { x: number; y: number } | null {
    if (!this.nvlInstance) return null;

    if (isRelationship) {
      const rels = this.nvlInstance.getRelationships();
      const rel = rels.find((r) => r.id === elementId);
      if (!rel) return null;
      const fromPos = this.nvlInstance.getPositionById(rel.from);
      const toPos = this.nvlInstance.getPositionById(rel.to);
      if (!fromPos || !toPos) return null;
      return {
        x: ((fromPos.x ?? 0) + (toPos.x ?? 0)) / 2,
        y: ((fromPos.y ?? 0) + (toPos.y ?? 0)) / 2,
      };
    } else {
      const pos = this.nvlInstance.getPositionById(elementId);
      if (!pos) return null;
      return { x: pos.x ?? 0, y: pos.y ?? 0 };
    }
  }

  /**
   * Converts an element's world position to CSS-pixel coordinates relative to
   * the `<nvl-graph>` host element.
   *
   * NVL's `getPan()` returns the camera/viewport offset in world space, so it
   * must be subtracted: screenPos = center + (worldPos - pan) * scale
   */
  private getElementScreenPosition(
    elementId: string,
    isRelationship: boolean
  ): { x: number; y: number } | null {
    if (!this.nvlInstance || !this.container) return null;

    const world = this.getWorldPosition(elementId, isRelationship);
    if (!world) return null;

    const scale = this.nvlInstance.getScale();
    const pan = this.nvlInstance.getPan();
    const thisRect = this.getBoundingClientRect();

    // Canvas renderer
    const canvas = this.container.querySelector("canvas");
    if (canvas) {
      const canvasRect = canvas.getBoundingClientRect();
      return {
        x:
          canvasRect.left -
          thisRect.left +
          canvasRect.width / 2 +
          (world.x - pan.x) * scale,
        y:
          canvasRect.top -
          thisRect.top +
          canvasRect.height / 2 +
          (world.y - pan.y) * scale,
      };
    }

    // SVG renderer
    const svg = this.container.querySelector("svg");
    if (svg) {
      const svgRect = svg.getBoundingClientRect();
      return {
        x:
          svgRect.left -
          thisRect.left +
          svgRect.width / 2 +
          (world.x - pan.x) * scale,
        y:
          svgRect.top -
          thisRect.top +
          svgRect.height / 2 +
          (world.y - pan.y) * scale,
      };
    }

    return null;
  }

  private updatePopupPosition() {
    if (!this.popupOverlay || !this.popupTargetId) return;

    const pos = this.getElementScreenPosition(
      this.popupTargetId,
      this.popupTargetIsRelationship
    );
    if (!pos) return;

    // Position the popup so it's centered horizontally above the element
    this.popupOverlay.style.left = `${pos.x}px`;
    this.popupOverlay.style.top = `${pos.y}px`;
    this.popupOverlay.style.transform = "translate(-50%, -100%)";
  }

  private startPopupTracking() {
    this.stopPopupTracking();
    const track = () => {
      this.updatePopupPosition();
      this.popupTrackingRaf = requestAnimationFrame(track);
    };
    this.popupTrackingRaf = requestAnimationFrame(track);
  }

  private stopPopupTracking() {
    if (this.popupTrackingRaf !== null) {
      cancelAnimationFrame(this.popupTrackingRaf);
      this.popupTrackingRaf = null;
    }
  }

  // ===== Lifecycle =====

  private destroyNvl() {
    this.stopPopupTracking();
    this.clickInteraction?.destroy();
    this.clickInteraction = null;
    this.dragNodeInteraction?.destroy();
    this.dragNodeInteraction = null;
    this.panInteraction?.destroy();
    this.panInteraction = null;
    this.zoomInteraction?.destroy();
    this.zoomInteraction = null;
    if (this.nvlInstance) {
      this.nvlInstance.destroy();
      this.nvlInstance = null;
    }
  }

  disconnectedCallback() {
    super.disconnectedCallback();
    this.destroyNvl();
  }
}

File: vaadin-nvl/src/main/kotlin/org/antoined/vaadin/nvl/NvlGraph.kt
package org.antoined.vaadin.nvl

import com.vaadin.flow.component.Component
import com.vaadin.flow.component.ComponentEvent
import com.vaadin.flow.component.ComponentEventListener
import com.vaadin.flow.component.DomEvent
import com.vaadin.flow.component.EventData
import com.vaadin.flow.component.HasSize
import com.vaadin.flow.component.Tag
import com.vaadin.flow.component.dependency.JsModule
import com.vaadin.flow.component.dependency.NpmPackage
import com.vaadin.flow.dom.DomListenerRegistration
import com.vaadin.flow.dom.Element
import com.vaadin.flow.shared.Registration
import elemental.json.Json
import elemental.json.JsonArray
import elemental.json.JsonObject
import elemental.json.JsonValue
import java.io.Serializable

/**
 * Vaadin Flow server-side component wrapping the
 * [Neo4j NVL](https://neo4j.com/docs/nvl/current/) JavaScript library for
 * interactive graph visualization.
 *
 * The component communicates with its client-side `nvl-graph.ts` web component
 * via properties, JS function calls, and DOM events.
 *
 * @param options Optional initial [NvlOptions] applied when the component is first attached.
 */
@Tag("nvl-graph")
@JsModule("./nvl-graph.ts")
@NpmPackage(value = "@neo4j-nvl/base", version = "1.1.0")
@NpmPackage(value = "@neo4j-nvl/interaction-handlers", version = "1.1.0")
class NvlGraph(options: NvlOptions? = null) : Component(), HasSize {

    init {
        options?.let { element.setPropertyJson("options", it.toJson()) }
    }

    // ===== Graph data =====

    /** Replaces all nodes in the graph. */
    fun setNodes(nodes: List<NvlNode>) {
        element.setPropertyJson("nodes", nodes.map { it.toJson() }.toJsonArray())
    }

    /** Replaces all relationships in the graph. */
    fun setRelationships(relationships: List<NvlRelationship>) {
        element.setPropertyJson("relationships", relationships.map { it.toJson() }.toJsonArray())
    }

    /** Convenience method to replace both nodes and relationships at once. */
    fun setGraph(nodes: List<NvlNode>, relationships: List<NvlRelationship>) {
        setNodes(nodes)
        setRelationships(relationships)
    }

    /** Adds new elements or updates existing ones (matched by id). */
    fun addAndUpdateElements(
        nodes: List<NvlNode> = emptyList(),
        relationships: List<NvlRelationship> = emptyList(),
    ) {
        element.callJsFunction(
            "addAndUpdateElements",
            nodes.map { it.toJson() }.toJsonArray(),
            relationships.map { it.toJson() }.toJsonArray(),
        )
    }

    /** Updates properties of existing nodes and/or relationships (matched by id). */
    fun updateElements(
        nodes: List<NvlNode> = emptyList(),
        relationships: List<NvlRelationship> = emptyList(),
    ) {
        element.callJsFunction(
            "updateElements",
            nodes.map { it.toJson() }.toJsonArray(),
            relationships.map { it.toJson() }.toJsonArray(),
        )
    }

    /** Adds new nodes and/or relationships to the graph. */
    fun addElements(
        nodes: List<NvlNode> = emptyList(),
        relationships: List<NvlRelationship> = emptyList(),
    ) {
        element.callJsFunction(
            "addElements",
            nodes.map { it.toJson() }.toJsonArray(),
            relationships.map { it.toJson() }.toJsonArray(),
        )
    }

    /** Removes nodes by their IDs; associated relationships are also removed. */
    fun removeNodes(nodeIds: List<String>) {
        element.callJsFunction("removeNodes", nodeIds.toJsonStringArray())
    }

    /** Removes relationships by their IDs. */
    fun removeRelationships(relationshipIds: List<String>) {
        element.callJsFunction("removeRelationships", relationshipIds.toJsonStringArray())
    }

    // ===== Layout =====

    /** Switches the graph to the given [layout] algorithm and restarts the layout. */
    fun setLayout(layout: NvlLayout) {
        element.callJsFunction("setLayout", layout.jsValue)
    }

    /** Sets algorithm-specific layout options (e.g. [NvlHierarchicalOptions]). */
    fun setLayoutOptions(options: NvlLayoutOptions) {
        element.callJsFunction("setLayoutOptions", options.toJson())
    }

    // ===== Zoom & pan =====

    /** Sets the zoom level to the given absolute value. */
    fun setZoom(zoomValue: Double) {
        element.callJsFunction("setZoom", zoomValue)
    }

    /** Sets the pan offset to the given pixel coordinates. */
    fun setPan(panX: Double, panY: Double) {
        element.callJsFunction("setPan", panX, panY)
    }

    /** Sets both zoom level and pan offset in a single call. */
    fun setZoomAndPan(zoom: Double, panX: Double, panY: Double) {
        element.callJsFunction("setZoomAndPan", zoom, panX, panY)
    }

    /** Resets the zoom level to its initial/default value. */
    fun resetZoom() {
        element.callJsFunction("resetZoom")
    }

    /** Adjusts zoom and pan to fit the given nodes into the viewport. */
    fun fit(nodeIds: List<String>, zoomOptions: NvlZoomOptions? = null) {
        if (zoomOptions != null) {
            element.callJsFunction("fit", nodeIds.toJsonStringArray(), zoomOptions.toJson())
        } else {
            element.callJsFunction("fit", nodeIds.toJsonStringArray())
        }
    }

    /** Adjusts zoom and pan to fit the entire graph into the viewport. */
    fun fitAll(zoomOptions: NvlZoomOptions? = null) {
        if (zoomOptions != null) {
            element.callJsFunction("fitAll", zoomOptions.toJson())
        } else {
            element.callJsFunction("fitAll")
        }
    }

    // ===== Selection =====

    /** Clears the selection state of all nodes and relationships. */
    fun deselectAll() {
        element.callJsFunction("deselectAll")
    }

    // ===== Pinning =====

    /** Pins a node so that layout forces no longer affect its position. */
    fun pinNode(nodeId: String) {
        element.callJsFunction("pinNode", nodeId)
    }

    /** Unpins nodes so that layout forces can move them again. */
    fun unpinNodes(nodeIds: List<String>) {
        element.callJsFunction("unpinNodes", nodeIds.toJsonStringArray())
    }

    // ===== Node positions =====

    /**
     * Moves nodes to explicit positions. Only [NvlNode.x] and [NvlNode.y] are used.
     *
     * @param updateLayout When `true`, the layout engine is informed of the new positions.
     */
    fun setNodePositions(nodes: List<NvlNode>, updateLayout: Boolean = false) {
        element.callJsFunction(
            "setNodePositions",
            nodes.map { it.toJson() }.toJsonArray(),
            updateLayout,
        )
    }

    // ===== Node dragging =====

    /** Enables or disables interactive node dragging by the user. */
    fun setNodeDraggingEnabled(enabled: Boolean) {
        element.setProperty("nodeDraggingEnabled", enabled)
    }

    // ===== Renderer =====

    /** Switches the rendering backend at runtime. */
    fun setRenderer(renderer: NvlRenderer) {
        element.callJsFunction("setRenderer", renderer.jsValue)
    }

    // ===== Restart =====

    /**
     * Restarts the NVL instance, optionally applying new [options].
     *
     * @param retainPositions When `true`, existing node positions are preserved across the restart.
     */
    fun restart(options: NvlOptions? = null, retainPositions: Boolean = false) {
        if (options != null) {
            element.callJsFunction("restart", options.toJson(), retainPositions)
        } else {
            element.callJsFunction("restart", Json.createObject(), retainPositions)
        }
    }

    // ===== Export =====

    /** Triggers a client-side PNG download of the visible graph area. */
    fun saveToFile(filename: String? = null, backgroundColor: String? = null) {
        element.callJsFunction("saveToFile", optionsJsonOf(filename, backgroundColor))
    }

    /** Triggers a client-side SVG download of the visible graph area. */
    fun saveToSvg(filename: String? = null, backgroundColor: String? = null) {
        element.callJsFunction("saveToSvg", optionsJsonOf(filename, backgroundColor))
    }

    /** Triggers a client-side PNG download of the full graph (may be larger than the viewport). */
    fun saveFullGraphToLargeFile(filename: String? = null, backgroundColor: String? = null) {
        element.callJsFunction("saveFullGraphToLargeFile", optionsJsonOf(filename, backgroundColor))
    }

    // ===== Client→Server getters (async) =====

    /** Asynchronously retrieves the current zoom scale from the client. */
    fun getScale(callback: SerializableCallback<Double>) {
        element.callJsFunction("getScale").then(Double::class.java) { callback.accept(it) }
    }

    /** Asynchronously retrieves the current pan offset from the client. */
    fun getPan(callback: SerializableCallback<NvlPoint>) {
        element.callJsFunction("getPan").then(JsonValue::class.java) { json ->
            val obj = json as JsonObject
            callback.accept(NvlPoint(obj.getNumber("x"), obj.getNumber("y")))
        }
    }

    /** Asynchronously retrieves the min/max zoom limits from the client. */
    fun getZoomLimits(callback: SerializableCallback<NvlZoomLimits>) {
        element.callJsFunction("getZoomLimits").then(JsonValue::class.java) { json ->
            val obj = json as JsonObject
            callback.accept(NvlZoomLimits(obj.getNumber("minZoom"), obj.getNumber("maxZoom")))
        }
    }

    /** Asynchronously retrieves all nodes currently in the graph from the client. */
    fun getNodes(callback: SerializableCallback<List<NvlNode>>) {
        element.callJsFunction("getNodes").then(JsonValue::class.java) { json ->
            callback.accept(parseNodeArray(json as JsonArray))
        }
    }

    /** Asynchronously retrieves all relationships currently in the graph from the client. */
    fun getRelationships(callback: SerializableCallback<List<NvlRelationship>>) {
        element.callJsFunction("getRelationships").then(JsonValue::class.java) { json ->
            callback.accept(parseRelationshipArray(json as JsonArray))
        }
    }

    /** Asynchronously retrieves the currently selected nodes from the client. */
    fun getSelectedNodes(callback: SerializableCallback<List<NvlNode>>) {
        element.callJsFunction("getSelectedNodes").then(JsonValue::class.java) { json ->
            callback.accept(parseNodeArray(json as JsonArray))
        }
    }

    /** Asynchronously retrieves the currently selected relationships from the client. */
    fun getSelectedRelationships(callback: SerializableCallback<List<NvlRelationship>>) {
        element.callJsFunction("getSelectedRelationships").then(JsonValue::class.java) { json ->
            callback.accept(parseRelationshipArray(json as JsonArray))
        }
    }

    /** Asynchronously retrieves the current x/y positions of all nodes from the client. */
    fun getNodePositions(callback: SerializableCallback<List<NvlNode>>) {
        element.callJsFunction("getNodePositions").then(JsonValue::class.java) { json ->
            callback.accept(parseNodeArray(json as JsonArray))
        }
    }

    /** Asynchronously checks whether the layout is still computing/animating. */
    fun isLayoutMoving(callback: SerializableCallback<Boolean>) {
        element.callJsFunction("isLayoutMoving").then(Boolean::class.java) { callback.accept(it) }
    }

    /** Asynchronously retrieves a Base64 data URL of the current graph rendering. */
    fun getImageDataUrl(callback: SerializableCallback<String>, backgroundColor: String? = null) {
        val opts = Json.createObject().apply {
            backgroundColor?.let { put("backgroundColor", it) }
        }
        element.callJsFunction("getImageDataUrl", opts).then(String::class.java) { callback.accept(it) }
    }

    // ===== Events =====

    /** Registers a listener for single-click events on nodes, relationships, or the canvas. */
    fun addClickListener(listener: ComponentEventListener<NvlClickEvent>): Registration =
        addListener(NvlClickEvent::class.java, listener)

    /** Registers a listener for double-click events on nodes or relationships. */
    fun addDoubleClickListener(listener: ComponentEventListener<NvlDoubleClickEvent>): Registration =
        addListener(NvlDoubleClickEvent::class.java, listener)

    /** Registers a listener for right-click (context menu) events on nodes or relationships. */
    fun addContextMenuListener(listener: ComponentEventListener<NvlContextMenuEvent>): Registration =
        addListener(NvlContextMenuEvent::class.java, listener)

    /** Registers a listener fired when the layout algorithm finishes computing. */
    fun addLayoutDoneListener(listener: ComponentEventListener<NvlLayoutDoneEvent>): Registration =
        addListener(NvlLayoutDoneEvent::class.java, listener)

    /** Registers a listener fired when the layout starts or stops computing. */
    fun addLayoutComputingListener(listener: ComponentEventListener<NvlLayoutComputingEvent>): Registration =
        addListener(NvlLayoutComputingEvent::class.java, listener)

    /** Registers a listener fired when a zoom/pan animation completes. */
    fun addZoomTransitionDoneListener(listener: ComponentEventListener<NvlZoomTransitionDoneEvent>): Registration =
        addListener(NvlZoomTransitionDoneEvent::class.java, listener)

    /** Registers a listener fired once the NVL instance is fully initialized on the client. */
    fun addInitializationListener(listener: ComponentEventListener<NvlInitializationEvent>): Registration =
        addListener(NvlInitializationEvent::class.java, listener)

    /** Registers a listener fired when the user finishes dragging a node. */
    fun addNodeDragEndListener(listener: ComponentEventListener<NvlNodeDragEndEvent>): Registration =
        addListener(NvlNodeDragEndEvent::class.java, listener)

    // ===== Event classes =====

    /** Fired on a single click. Carries the IDs of hit nodes and relationships. */
    @DomEvent("nvl-click")
    class NvlClickEvent(
        source: NvlGraph,
        fromClient: Boolean,
        @EventData("event.detail.nodeIds") nodeIds: JsonArray?,
        @EventData("event.detail.relationshipIds") relationshipIds: JsonArray?,
    ) : NvlHitEvent(source, fromClient, nodeIds, relationshipIds)

    /** Fired on a double click. Carries the IDs of hit nodes and relationships. */
    @DomEvent("nvl-dblclick")
    class NvlDoubleClickEvent(
        source: NvlGraph,
        fromClient: Boolean,
        @EventData("event.detail.nodeIds") nodeIds: JsonArray?,
        @EventData("event.detail.relationshipIds") relationshipIds: JsonArray?,
    ) : NvlHitEvent(source, fromClient, nodeIds, relationshipIds)

    /** Fired on a right-click / context menu. Carries the IDs of hit nodes and relationships. */
    @DomEvent("nvl-contextmenu")
    class NvlContextMenuEvent(
        source: NvlGraph,
        fromClient: Boolean,
        @EventData("event.detail.nodeIds") nodeIds: JsonArray?,
        @EventData("event.detail.relationshipIds") relationshipIds: JsonArray?,
    ) : NvlHitEvent(source, fromClient, nodeIds, relationshipIds)

    /**
     * Base class for pointer events that may hit nodes and/or relationships.
     *
     * @property nodeIds IDs of nodes under the pointer (empty if none).
     * @property relationshipIds IDs of relationships under the pointer (empty if none).
     */
    abstract class NvlHitEvent(
        source: NvlGraph,
        fromClient: Boolean,
        nodeIds: JsonArray?,
        relationshipIds: JsonArray?,
    ) : ComponentEvent<NvlGraph>(source, fromClient) {
        val nodeIds: List<String> = nodeIds?.toStringList() ?: emptyList()
        val relationshipIds: List<String> = relationshipIds?.toStringList() ?: emptyList()
    }

    /** Fired when the layout algorithm completes. */
    @DomEvent("nvl-layout-done")
    class NvlLayoutDoneEvent(
        source: NvlGraph,
        fromClient: Boolean,
    ) : ComponentEvent<NvlGraph>(source, fromClient)

    /** Fired when the layout computing state changes. Check [isComputing] for the current state. */
    @DomEvent("nvl-layout-computing")
    class NvlLayoutComputingEvent(
        source: NvlGraph,
        fromClient: Boolean,
        @EventData("event.detail.isComputing") val isComputing: Boolean,
    ) : ComponentEvent<NvlGraph>(source, fromClient)

    /** Fired when a zoom or fit animation completes. */
    @DomEvent("nvl-zoom-transition-done")
    class NvlZoomTransitionDoneEvent(
        source: NvlGraph,
        fromClient: Boolean,
    ) : ComponentEvent<NvlGraph>(source, fromClient)

    /** Fired once the NVL client-side instance is fully initialized and ready. */
    @DomEvent("nvl-initialization")
    class NvlInitializationEvent(
        source: NvlGraph,
        fromClient: Boolean,
    ) : ComponentEvent<NvlGraph>(source, fromClient)

    /**
     * Fired when the user finishes dragging a node.
     *
     * @property nodeId ID of the dragged node.
     * @property x Final horizontal position.
     * @property y Final vertical position.
     */
    @DomEvent("nvl-node-drag-end")
    class NvlNodeDragEndEvent(
        source: NvlGraph,
        fromClient: Boolean,
        @EventData("event.detail.nodeId") val nodeId: String,
        @EventData("event.detail.x") val x: Double,
        @EventData("event.detail.y") val y: Double,
    ) : ComponentEvent<NvlGraph>(source, fromClient)

    // ===== Helpers =====

    private fun optionsJsonOf(filename: String?, backgroundColor: String?): JsonObject =
        Json.createObject().apply {
            filename?.let { put("filename", it) }
            backgroundColor?.let { put("backgroundColor", it) }
        }

    private fun parseNodeArray(arr: JsonArray): List<NvlNode> =
        (0 until arr.length()).map { NvlNode.fromJson(arr.getObject(it)) }

    private fun parseRelationshipArray(arr: JsonArray): List<NvlRelationship> =
        (0 until arr.length()).map { NvlRelationship.fromJson(arr.getObject(it)) }
}

/**
 * A serializable callback for receiving asynchronous results from client-side JS calls.
 *
 * @param T The type of value returned by the client.
 */
fun interface SerializableCallback<T> : Serializable {
    fun accept(value: T)
}

/** A 2D point representing a pan offset or position in graph coordinates. */
data class NvlPoint(val x: Double, val y: Double)

/** The effective minimum and maximum zoom levels as reported by the client. */
data class NvlZoomLimits(val minZoom: Double, val maxZoom: Double)
